### PR TITLE
[KYUUBI #1832] Fixed: forcedMaxOutputRows extension for subquery

### LINKED ISSUE #1832: [Bug] Using the forcedMaxOutputRows extension for statements with subqueries can cause problems with the number of results returned
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe

Statements with subqueries have the limit condition error pushed down for the final generated logicalplan after setting watchdog.forcedMaxOutputRows.

SQL: 
```
CREATE TABLE spark_catalog.`default`.tmp_table1(KEY INT, VALUE STRING) USING PARQUET;
INSERT INTO TABLE spark_catalog.`default`.tmp_table1 VALUES (1, 'aa'),(2,'bb'),(3, 'cc'),(4,'aa'),(5,'cc'),(6, 'aa');

select count(*)
from tmp_table1
where tmp_table1.key in (
  select distinct tmp_table1.key
  from tmp_table1
  where tmp_table1.value = "aa"
  );
```
Analyzed LogicPlan: 
```
Aggregate [count(1) AS count(1)#62L]
+- Filter key#56 IN (list#60 [])
   :  +- GlobalLimit 1
   :     +- LocalLimit 1
   :        +- Distinct
   :           +- Project [key#56]
   :              +- Filter (value#57 = aa)
   :                 +- SubqueryAlias spark_catalog.default.tmp_table1
   :                    +- Relation[KEY#56,VALUE#57] parquet
   +- SubqueryAlias spark_catalog.default.tmp_table1
      +- Relation[KEY#56,VALUE#57] parquet
```
Limit rules are pushed down into the Filter, which will result in an inaccurate number of final query results.

### Bug Tracing
The `extensions.injectPostHocResolutionRule(ForcedMaxOutputRowsRule)` rule is injected in the analyzer phase, and a look at the Batches rule in the analyzer phase reveals that: 
image
SQL with subqueries will first enter the ResolveSubquery Batch during analyze, and the implementation of `ResolvedSubQuery` can be known:
image
The Analyzer's Execute method will be called again, and the analysis of SQL statements with subqueries will recursively call Analyzer's Batches. Since we added the `ForcedMaxOutputRowsRule`, the subqueries will also be limited with Limit Rule, which eventually leads to the above generated logical plan's semantics being inconsistent with what we expect.

### Solutions
1. Place the `ForcedMaxOutputRowsRule` in the `projectOptimizerRule` stage to avoid recursive subquery calls.
2. Remove `extensions.injectResolutionRule(MarkAggregateOrderRule)`. 
###  Some questions: 
I don't see any relevant cases in the unit tests that hit the Aggregate rule without the Limit restriction. And after I removed the markAgg exetension and put the `ForcedMaxOutputRowsRule` in the projectOptimizerRule phase, it passed all the unit tests

### Affects Version(s)

master/1.4.0

### Kyuubi Server Log Output

_No response_

### Kyuubi Engine Log Output

_No response_

### Kyuubi Server Configurations

_No response_

### Kyuubi Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by iodone**
### _Why are the changes needed?_

Fixed #1832 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/KyuubiSparkSQLExtension.scala [hunk 1]
```diff
@@ -40,8 +40,9 @@ class KyuubiSparkSQLExtension extends (SparkSessionExtensions => Unit) {
 
     // watchdog extension
     // a help rule for ForcedMaxOutputRowsRule
-    extensions.injectResolutionRule(MarkAggregateOrderRule)
-    extensions.injectPostHocResolutionRule(ForcedMaxOutputRowsRule)
+//    extensions.injectResolutionRule(MarkAggregateOrderRule)
+//    extensions.injectPostHocResolutionRule(ForcedMaxOutputRowsRule)
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/KyuubiSparkSQLExtension.scala:44]: we can just remove line 42 ~ 44

File: dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsBase.scala [hunk 2]
```diff
@@ -69,6 +69,7 @@ trait ForcedMaxOutputRowsBase extends Rule[LogicalPlan] {
       } else {
         true
       }
+    case _: MultiInstanceRelation => true
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsBase.scala:72]: why change this ?

Comment (iodone) [reply dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsBase.scala:72]: Place the ForcedMaxOutputRowsRule in the projectOptimizerRule stage，SQL like this: 
```
select * from table_a
```
this logicalPlan will be optimized as a `Relation` instead of a `Project`

Comment (ulysses-you) [reply dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsBase.scala:72]: I see the project here is been removed since the output is same

Comment (iodone) [reply dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsBase.scala:72]: so we need to cover this situation with `case _: MultiInstanceRelation => true` ?

Comment (ulysses-you) [reply dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsBase.scala:72]: I think it's fine if all test pass

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/KyuubiSparkSQLExtension.scala [hunk 3]
```diff
@@ -39,9 +39,7 @@ class KyuubiSparkSQLExtension extends (SparkSessionExtensions => Unit) {
     extensions.injectPostHocResolutionRule(DropIgnoreNonexistent)
 
     // watchdog extension
-    // a help rule for ForcedMaxOutputRowsRule
-    extensions.injectResolutionRule(MarkAggregateOrderRule)
```
Comment (cfmcgrady) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/KyuubiSparkSQLExtension.scala:43]: Since we have removed `extensions.injectResolutionRule(MarkAggregateOrderRule)`, it's safety to remove the case class `MarkAggregateOrderRule` defined in the file `dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/watchdog/ForcedMaxOutputRowsRule.scala`.

File: dev/kyuubi-extension-spark-common/src/test/scala/org/apache/spark/sql/WatchDogSuiteBase.scala [hunk 4]
```diff
@@ -411,8 +412,38 @@ trait WatchDogSuiteBase extends KyuubiSparkSQLExtensionTest {
...
+          """
+            |select count(*)
+            |from tmp_table1
+            |where tmp_table1.key in (
+            |select distinct tmp_table1.key
+            |from tmp_table1
+            |where tmp_table1.value = "aa"
+            |)
+            |""".stripMargin).queryExecution.optimizedPlan
+        assert(!findGlobalLimit(plan))
```
Comment (cfmcgrady) [inline dev/kyuubi-extension-spark-common/src/test/scala/org/apache/spark/sql/WatchDogSuiteBase.scala:438]: We'd better verify the results
```scala
    val q = sql(
      """
        |select count(*)
        |from tmp_table1
        |where tmp_table1.key in (
        |select distinct tmp_table1.key
        |from tmp_table1
        |where tmp_table1.value = "aa"
        |)
        |""".stripMargin)
    checkAnswer(q, Row(3) :: Nil)
```

Comment (iodone) [reply dev/kyuubi-extension-spark-common/src/test/scala/org/apache/spark/sql/WatchDogSuiteBase.scala:438]: Ok, I will add this case

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Comment by ulysses-you** [comment]
@iodone there has some conflicts with branch-1.4, can you send a backport ? thank you !

**Comment by iodone** [comment]
Let me try
